bazel: require cgroup sandboxing in CI

rockwotj opened the pull request:
This is required so we can run tests that use the local network in parallel. We can check it's working in our tools/bazel script, which runs as a hook before bazel is invoked.

From the bazel documentation (we want linux sandboxing enabled):

>linux-sandbox goes one step further and builds on top of the processwrapper-sandbox. Similar to what Docker does under the hood, it uses Linux Namespaces (User, Mount, PID, Network and IPC namespaces) to isolate the action from the host. That is, it makes the entire filesystem read-only except for the sandbox directory, so the action cannot accidentally modify anything on the host filesystem. This prevents situations like a buggy test accidentally rm -rf'ing your $HOME directory. Optionally, you can also prevent the action from accessing the network. linux-sandbox uses PID namespaces to prevent the action from seeing any other processes and to reliably kill all processes (even daemons spawned by the action) at the end.

>Both the linux-sandbox and the darwin-sandbox do not work in a "nested" scenario due to restrictions in the mechanisms provided by the operating systems. Because Docker also uses Linux namespaces for its container magic, you cannot easily run linux-sandbox inside a Docker container, unless you use docker run --privileged. On macOS, you cannot run sandbox-exec inside a process that's already being sandboxed. Thus, in these cases, Bazel automatically falls back to using processwrapper-sandbox.


We want to not implicitly fallback, but some tests require no sandboxing (ie. //:gazelle_test). However, in order to remove the exclusive tag for these tests, we need to enforce the highest form of sandboxing is used, otherwise we could run into port reuse issues. So add a check to ensure that we properly are using that form of sandboxing.

The check is done in the same way Bazel itself checks if the sandbox is functioning, by ensuring the true command exits successfully in the sandbox (the daemon caches this result).


## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

rockwotj committed:
bazel: require cgroup sandboxing in CI

This is required so we can run tests that use the local network in parallel. We can check it's working in our tools/bazel script, which runs as a hook before bazel is invoked.

From the bazel documentation (we want linux sandboxing enabled):

```
linux-sandbox goes one step further and builds on top of the processwrapper-sandbox. Similar to what Docker does under the hood, it uses Linux Namespaces (User, Mount, PID, Network and IPC namespaces) to isolate the action from the host. That is, it makes the entire filesystem read-only except for the sandbox directory, so the action cannot accidentally modify anything on the host filesystem. This prevents situations like a buggy test accidentally rm -rf'ing your $HOME directory. Optionally, you can also prevent the action from accessing the network. linux-sandbox uses PID namespaces to prevent the action from seeing any other processes and to reliably kill all processes (even daemons spawned by the action) at the end.

Both the linux-sandbox and the darwin-sandbox do not work in a "nested" scenario due to restrictions in the mechanisms provided by the operating systems. Because Docker also uses Linux namespaces for its container magic, you cannot easily run linux-sandbox inside a Docker container, unless you use docker run --privileged. On macOS, you cannot run sandbox-exec inside a process that's already being sandboxed. Thus, in these cases, Bazel automatically falls back to using processwrapper-sandbox.
```

We want to not implicitly fallback, but some tests require no sandboxing (ie. //:gazelle_test). Because we run in a docker container for CI, we want to ensure sandboxing is actually enabled.

The check is done in the same way Bazel itself checks if the sandbox is functioning, by ensuring the true command exits successfully in the sandbox (the daemon caches this result).

rockwotj commented:
Since this passed CI - I believe this means we can remove the `exclusive` tags. Will do some more testing first

rockwotj committed:
treewide: remove exclusive test tags

Because we run CI with linux-sandboxing we get a cgroup per test,
meaning we don't have to worry about tests having conflicting ports or
data directories.

rockwotj committed:
treewide: use bazel tmp directory to write data too

See: https://bazel.build/reference/test-encyclopedia#test-interaction-filesystem

It's not valid to write to CWD and we break test concurrency in doing
so.

This should probably not copy/paste this helper all over, but this is my
lazy way of checking if this makes it work.

rockwotj committed:
bazel: check for CI=true

We really only want/need this in buildkite - we don't strict sandboxing
in lint-cpp for github actions.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65047-->
<!--974f68d2-90c9-4462-8c7a-657beaa45c89-->
<!--title-->
## Retry command for [Build#65047](https://buildkite.com/redpanda/redpanda/builds/65047)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":true,"mixed_versions":false,"with_chunked_compaction":true,"with_iceberg":true,"with_tiered_storage":false}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--e249843a-969b-42be-9496-a59ece1a4171-->
<!--TEST RESULTS-->
<!--a85c2d01-e72f-46c1-98ec-40028c18462f-->
<!--TEST RESULTS-->
<!--25bef9c0-0d0c-4bdb-9f6b-98a6518fa5e2-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65047">build#65047</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_debug_bundle_service_rpunit.gtest_debug_bundle_service_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65047#01966912-2618-48fb-9b26-2d49bb91d48f | FAIL | 0/2 |
| gtest_debug_bundle_service_rpunit.gtest_debug_bundle_service_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65047#01966912-2619-4daf-959f-b71f89645696 | FLAKY | 1/2 |
| idempotency_tests_rpunit.idempotency_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65047#01966912-2618-48fb-9b26-2d49bb91d48f | FAIL | 0/2 |
| idempotency_tests_rpunit.idempotency_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65047#01966912-2619-4daf-959f-b71f89645696 | FAIL | 0/2 |
| rm_stm_tests_rpunit.rm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65047#01966912-2618-48fb-9b26-2d49bb91d48f | FAIL | 0/2 |
| rm_stm_tests_rpunit.rm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65047#01966912-2619-4daf-959f-b71f89645696 | FAIL | 0/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/65047#0196696a-649e-4e57-872b-662cc174dd1d | FLAKY | 18/21 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/65047#0196695c-4282-4021-933c-9deadc27243c | FLAKY | 20/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/65047#0196696a-649d-4c36-8a87-d5dc6db7646e | FLAKY | 15/21 |
| tx_compaction_tests_rpunit.tx_compaction_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65047#01966912-2618-48fb-9b26-2d49bb91d48f | FAIL | 0/2 |
| tx_compaction_tests_rpunit.tx_compaction_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65047#01966912-2619-4daf-959f-b71f89645696 | FAIL | 0/2 |
| utils_single_thread_rpunit.utils_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65047#01966912-2618-48fb-9b26-2d49bb91d48f | FAIL | 0/2 |
| utils_single_thread_rpunit.utils_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65047#01966912-2619-4daf-959f-b71f89645696 | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65188">build#65188</a></summary>


| test_id                                                                                                                      | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_debug_bundle_service_rpunit.gtest_debug_bundle_service_rpunit                                                          | unit        | https://buildkite.com/redpanda/redpanda/builds/65188#01967ce1-be83-4f8f-bc00-521403ca9bfd | FAIL          | 0/2      |
| gtest_debug_bundle_service_rpunit.gtest_debug_bundle_service_rpunit                                                          | unit        | https://buildkite.com/redpanda/redpanda/builds/65188#01967ce1-be84-4fd5-827b-3497056db103 | FAIL          | 0/2      |
| idempotency_tests_rpunit.idempotency_tests_rpunit                                                                            | unit        | https://buildkite.com/redpanda/redpanda/builds/65188#01967ce1-be83-4f8f-bc00-521403ca9bfd | FAIL          | 0/2      |
| idempotency_tests_rpunit.idempotency_tests_rpunit                                                                            | unit        | https://buildkite.com/redpanda/redpanda/builds/65188#01967ce1-be84-4fd5-827b-3497056db103 | FAIL          | 0/2      |
| rm_stm_tests_rpunit.rm_stm_tests_rpunit                                                                                      | unit        | https://buildkite.com/redpanda/redpanda/builds/65188#01967ce1-be83-4f8f-bc00-521403ca9bfd | FAIL          | 0/2      |
| rm_stm_tests_rpunit.rm_stm_tests_rpunit                                                                                      | unit        | https://buildkite.com/redpanda/redpanda/builds/65188#01967ce1-be84-4fd5-827b-3497056db103 | FAIL          | 0/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65188#01967d22-33ea-4a64-ab15-6e47187e5e59 | FLAKY         | 7/21     |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65188#01967d3e-288b-4b0a-ba2a-dd0c41fb7c39 | FLAKY         | 3/21     |
| rptest.tests.datalake.iceberg_toggling_test.IcebergTogglingTest.test_iceberg_toggling.cloud_storage_type=CloudStorageType.S3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65188#01967d22-33e8-4a17-8179-76bc667ba9c1 | FLAKY         | 20/21    |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_add_partitions_with_inprogress_reassignments       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65188#01967d22-33e8-4a17-8179-76bc667ba9c1 | FLAKY         | 20/21    |
| rptest.tests.write_caching_fi_e2e_test.WriteCachingFailureInjectionE2ETest.test_crash_all_with_consumer_group                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65188#01967d22-33e9-4d70-9264-10e222dd0970 | FLAKY         | 20/21    |
| tx_compaction_tests_rpunit.tx_compaction_tests_rpunit                                                                        | unit        | https://buildkite.com/redpanda/redpanda/builds/65188#01967ce1-be83-4f8f-bc00-521403ca9bfd | FAIL          | 0/2      |
| tx_compaction_tests_rpunit.tx_compaction_tests_rpunit                                                                        | unit        | https://buildkite.com/redpanda/redpanda/builds/65188#01967ce1-be84-4fd5-827b-3497056db103 | FAIL          | 0/2      |
| utils_single_thread_rpunit.utils_single_thread_rpunit                                                                        | unit        | https://buildkite.com/redpanda/redpanda/builds/65188#01967ce1-be83-4f8f-bc00-521403ca9bfd | FAIL          | 0/2      |
| utils_single_thread_rpunit.utils_single_thread_rpunit                                                                        | unit        | https://buildkite.com/redpanda/redpanda/builds/65188#01967ce1-be84-4fd5-827b-3497056db103 | FAIL          | 0/2      |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65636">build#65636</a></summary>


| test_id                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65636#0196ab51-1ffc-40a5-8df9-51f7a3e0cb56 | FLAKY         | 9/21     |
</details>
<!--/body-->

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/4d645af0-7cdb-40a4-b01d-9c0fe685a72d?refs=30cbba286cf616a0edc51f9085e6b0535ddf6472&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/4d645af0-7cdb-40a4-b01d-9c0fe685a72d/license?refs=30cbba286cf616a0edc51f9085e6b0535ddf6472&source=prChecksComment)

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/97765103-bf91-4050-82d7-08c613cf32ef?refs=30cbba286cf616a0edc51f9085e6b0535ddf6472&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/97765103-bf91-4050-82d7-08c613cf32ef/license?refs=30cbba286cf616a0edc51f9085e6b0535ddf6472&source=prChecksComment)

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65188-->
<!--b96e00d4-b5ec-4af3-ad70-d57defb83e59-->
<!--0f1159be-eedf-4658-9c2e-00f1d59b79c2-->
<!--3afc3a55-1b46-43ce-9f93-7f3377c733ea-->
<!--ee2956bb-938a-46e5-a6ef-37ace2256670-->
<!--title-->
## Retry command for [Build#65188](https://buildkite.com/redpanda/redpanda/builds/65188)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names
tests/rptest/tests/write_caching_fi_e2e_test.py::WriteCachingFailureInjectionE2ETest.test_crash_all_with_consumer_group
tests/rptest/tests/partition_reassignments_test.py::PartitionReassignmentsTest.test_add_partitions_with_inprogress_reassignments
tests/rptest/tests/datalake/iceberg_toggling_test.py::IcebergTogglingTest.test_iceberg_toggling@{"cloud_storage_type":1}
```
<!--/body-->

rockwotj committed:
bazel/cloud_storage: fix manual fixture directory

It needs to be in a Bazel writeable directory.

rockwotj committed:
wasmtime: upgrade to latest version (v32)

This changes some of the APIs around memory management, although the end
result is still the same.

rockwotj commented:
This is ready for review now 🎉

rockwotj committed:
debug_bundle: deflake test

The file was being flushed async (why?), and reading the file was racing
with the truncation after the direct I/O write.

clee reviewed (approved):
LGTM

rockwotj commented:
> @rockwotj - so is "cgroup sandboxing" currently used for dev (non-CI) and this change enables it _also_ for CI, or it will be only in CI?

This change _requires_ it in CI. For local development, it depends if the strict standboxing using cgroups is enabled. It might require some settings changes on your local machine. I have a post in #team-core on how to do so, I will also add to the wiki.

You can test if the cgroup sandboxing is enabled locally using:

```
`bazel info install_base`/linux-sandbox /bin/true; echo $?
```

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65636-->
<!--5af3f866-ab25-4cb3-9dcb-3945f47eb95b-->
<!--title-->
## Retry command for [Build#65636](https://buildkite.com/redpanda/redpanda/builds/65636)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/upgrade_test.py::UpgradeBackToBackTest.test_upgrade_with_all_workloads@{"single_upgrade":false}
```
<!--/body-->

dotnwat commented:
> Because we run in a docker container for CI, we want to ensure sandboxing is actually enabled.

i didn't really understand this from the coverletter.

dotnwat commented on src/v/debug_bundle/debug_bundle_service.cc:
```diff
@@ -761,7 +751,7 @@ ss::future<> service::set_metadata(job_id_t job_id) {
 
     try {
-        co_await write_file(process_output_file.native(), std::move(po_buf));
+        co_await write_fully(process_output_file, std::move(po_buf));
```
> The file was being flushed async (why?), and reading the file was racing
with the truncation after the direct I/O write.

I'm not following this. From what I can tell, `write_file` was synchronously flushing (last line in the function). With respect to the race, what's the read path you were referring to?

rockwotj replied:
> last line in the function 

Ah that is not flushing the file but the stream!

rockwotj replied:
Oh wait maybe the stream flush does flush the file... Hrm, the race must be elsewhere

dotnwat replied:
yeh i think it does. output_stream::flush -> file_data_sink_impl::flush -> file::flush.

dotnwat replied:
@nvartolomei knows for sure i think he filed a ticket recently in seastar to change this behavior

rockwotj replied:
Okay with Mike's help I tracked down the proper fix: https://github.com/redpanda-data/redpanda/pull/26060

rockwotj commented:
> > Because we run in a docker container for CI, we want to ensure sandboxing is actually enabled.
> 
> i didn't really understand this from the coverletter.

Replaced that with the following:

However, in order to remove the exclusive tag for these tests, we need to enforce the highest form of sandboxing is used, otherwise we could run into port reuse issues. So add a check to ensure that we properly are using that form of sandboxing.

rockwotj commented:
> > This change _requires_ it in CI. For local development, it depends if the strict standboxing using cgroups is enabled. It might require some settings changes on your local machine. I have a post in #team-core on how to do so, I will also add to the wiki.
> > You can test if the cgroup sandboxing is enabled locally using:
> 
> Thanks. This explains a confusion I had as I noticed on a new machine I saw `bazel-processwrapper` for all my builds, whereas on older boxes it always said `linux-sandbox` which I now can guess it just due to the sanbox not being installed on my new box.

From slack, I think these will get the sandbox going for you (assuming ubuntu LTS)

Enable for a single boot:
```
echo 0 | sudo tee /proc/sys/kernel/apparmor_restrict_unprivileged_userns
```
Forever:
```
echo "kernel.apparmor_restrict_unprivileged_userns=0" | sudo tee /etc/sysctl.d/60-apparmor-namespace.conf
```

rockwotj commented:
> LIke maybe it's something we should put in bazel/install-deps.sh bootstrap script.

Not opposed to it, or even requiring it for environments outside of CI (basically removing the is CI check in `tools/bazel`). In practice there are a few things it can be so it might be hard to automated all the steps needed
